resource-managers/kubernetes/token-refresh-server/src/main/scala/org/apache/spark/security/kubernetes/TokenRefreshService.scala

```diff
@@ -0,0 +1,361 @@
+    val newToken = ugi.doAs(new PrivilegedExceptionAction[Token[_ <: TokenIdentifier]] {
+
+      override def run() : Token[_ <: TokenIdentifier] = {
+        val fs = FileSystem.get(hadoopConf)
+        val tokens = fs.addDelegationTokens(UserGroupInformation.getLoginUser.getUserName,
+          credentials)
+        tokens(0)
+      }
+    })
+    newToken
+  }
+
+  private def writeTokensToSecret(tokenToExpire: Map[Token[_ <: TokenIdentifier], Long],
+                                  nowMillis: Long) : Unit = {
+    val durationUntilExpire = tokenToExpire.values.min - nowMillis
+    val key = s"$SECRET_DATA_ITEM_KEY_PREFIX_HADOOP_TOKENS$nowMillis-$durationUntilExpire"
+    val credentials = new Credentials()
+    tokenToExpire.keys.foreach(token => credentials.addToken(token.getService, token))
+    val serialized = serializeCredentials(credentials)
+    val value = Base64.encodeBase64String(serialized)
+    val secretMeta = renew.secretMeta
+    val editor = kubernetesClient.secrets
+      .inNamespace(secretMeta.getNamespace)
+      .withName(secretMeta.getName)
+      .edit()
+    editor.addToData(key, value)
+    val dataItemKeys = editor.getData.keySet().asScala.filter(
+      _.startsWith(SECRET_DATA_ITEM_KEY_PREFIX_HADOOP_TOKENS)).toSeq.sorted
+    // Remove data items except the latest two data items. A K8s secret can hold only up to 1 MB
+    // data. We need to remove old data items. We keep the latest two items to avoid race conditions
+    // where some newly launching executors may access the previous token.
+    dataItemKeys.dropRight(2).foreach(editor.removeFromData)
+    editor.done
+    logInfo(s"Wrote new tokens $tokenToExpire to a data item $key in ${secretMeta.getSelfLink}")
+  }
+
+  private def serializeCredentials(credentials: Credentials) = {
+    val byteStream = new ByteArrayOutputStream
+    val dataStream = new DataOutputStream(byteStream)
+    credentials.writeTokenStorageToStream(dataStream)
+    dataStream.flush()
+    byteStream.toByteArray
+  }
+
+  private def getRetryTime = clock.nowInMillis() + RENEW_TASK_RETRY_TIME_MILLIS
+}
+
+private class Clock {
+
+  def nowInMillis() : Long = System.currentTimeMillis()
+}
+
+private sealed trait Command
+private case object Relogin extends Command
+private case class UpdateSecretsToTrack(secrets: List[Secret]) extends Command
+private case class StartRefresh(secret: Secret) extends Command
+private case class Renew(expireTime: Long,
+                         tokenToExpireTime: Map[Token[_ <: TokenIdentifier], Long],
+                         secretMeta: ObjectMeta,
+                         numConsecutiveErrors: Int) extends Command
+private case class StopRefresh(secret: Secret) extends Command
+
+private object TokenRefreshService {
+
+  val hadoopTokenPattern = Pattern.compile(SECRET_DATA_ITEM_KEY_REGEX_HADOOP_TOKENS)
+
+  def apply(system: ActorSystem, kubernetesClient: KubernetesClient) : ActorRef = {
+    UserGroupInformation.loginUserFromKeytab(
```

Comment:
How is this being used?

Reply:
`loginUsersFromKeytab` sets the login user of the application as a static singleton field. So it gets used later as the proxy user when a new job user token is obtained.